tls/ttls.c

```diff
@@ -2198,12 +2200,17 @@ ttls_recv(void *tls_data, unsigned char *buf, size_t len, unsigned int *read)
 
 next_record:
 	if (!(io->st_flags & TTLS_F_ST_HDRIV)) {
+		unsigned char old_hdr_cpsz = io->hdr_cpsz;
 		if ((r = ttls_parse_record_hdr(tls, buf, len, read)))
 			return r;
 		if (io->msgtype == TTLS_MSG_HANDSHAKE
 		    && ttls_hs_checksumable(tls))
 		{
-			if (likely(*read - parsed >= TTLS_HS_HDR_LEN)) {
+			if (likely(old_hdr_cpsz <= TLS_HEADER_SIZE &&
+			           old_hdr_cpsz + *read - parsed >
+			                   TLS_HEADER_SIZE + TTLS_HS_HDR_LEN
```

Comment:
I believe the second part is always false, and that effectively turns the checksumming trick off. Still, it should work that way, but we'll play safely then.

Instead, I'd like to propose this way:
```patch
+                       if (likely(*read - parsed >= TTLS_HS_HDR_LEN &&
+                                  len > *read - parsed))
```
We keep the first part of the condition, as it's true when we've just finished header parsing, and at least handshake header is in the continuous buffer. The second part is the addition. We need to have at least one more byte in the continuous buffer to postpone checksum calculation. Otherwise we exit in the `if (delta == len)` below, and never factor the handshake header into a checksum.